File: net/crates/net/sdk-ts/test/cortex.test.ts
// Integration tests for the CortEX SDK wrapper in sdk-ts/src/cortex.ts.
//
// Exercises the AsyncIterable-wrapped watch iterators and the
// snapshotAndWatch race-fix surface end-to-end through the napi
// boundary. Mirrors the Rust-side regression suite in
// `net/crates/net/tests/integration_cortex_tasks.rs` so a regression
// in either layer fails a test here.

import { describe, expect, it } from 'vitest';

import {
  CortexError,
  MemoriesAdapter,
  NetDb,
  Redex,
  TasksAdapter,
  TaskStatus,
  type Task,
} from '../src/cortex';

const ORIGIN = 0xabcdef01n;

function nowNs(): bigint {
  return BigInt(Date.now()) * 1_000_000n;
}

describe('NetDb', () => {
  it('opens with both models and exposes typed adapters', async () => {
    const db = await NetDb.open({
      originHash: ORIGIN,
      withTasks: true,
      withMemories: true,
    });
    expect(db.tasks).toBeInstanceOf(TasksAdapter);
    expect(db.memories).toBeInstanceOf(MemoriesAdapter);
    db.close();
  });

  it('opens with only tasks; memories is null', async () => {
    const db = await NetDb.open({ originHash: ORIGIN, withTasks: true });
    expect(db.tasks).toBeInstanceOf(TasksAdapter);
    expect(db.memories).toBeNull();
    db.close();
  });

  it('snapshot + restore round-trips', async () => {
    const db = await NetDb.open({
      originHash: ORIGIN,
      withTasks: true,
      withMemories: true,
    });
    const taskSeq = db.tasks!.create(1n, 'alpha', 100n);
    const memSeq = db.memories!.store(1n, 'mem', ['x'], 'alice', 100n);
    // Wait for BOTH fold tasks before snapshotting — otherwise the
    // tasks fold may not have applied the create yet and the restored
    // db would see count=0.
    await db.tasks!.waitForSeq(taskSeq);
    await db.memories!.waitForSeq(memSeq);

    const bundle = db.snapshot();
    const restored = await NetDb.openFromSnapshot(
      { originHash: ORIGIN, withTasks: true, withMemories: true },
      bundle,
    );
    expect(restored.tasks!.count()).toBe(1);
    expect(restored.memories!.count()).toBe(1);
  });
});

describe('TasksAdapter — CRUD + listTasks', () => {
  it('creates, renames, completes, deletes', async () => {
    const redex = new Redex();
    const tasks = await TasksAdapter.open(redex, ORIGIN);
    const t0 = nowNs();

    tasks.create(1n, 'a', t0);
    tasks.create(2n, 'b', t0 + 1n);
    tasks.rename(1n, 'a-renamed', t0 + 2n);
    const seq = tasks.complete(2n, t0 + 3n);
    await tasks.waitForSeq(seq);

    const all = tasks.listTasks(null);
    expect(all).toHaveLength(2);
    expect(all.find((t) => t.id === 1n)!.title).toBe('a-renamed');
    expect(all.find((t) => t.id === 2n)!.status).toBe(TaskStatus.Completed);
  });

  it('delete removes the task from the materialized state', async () => {
    const redex = new Redex();
    const tasks = await TasksAdapter.open(redex, ORIGIN);
    tasks.create(1n, 'tmp', nowNs());
    const seq = tasks.delete(1n);
    await tasks.waitForSeq(seq);
    expect(tasks.count()).toBe(0);
  });
});

describe('TasksAdapter.watch — for-await lifecycle', () => {
  it('emits initial result, then deltas, and close() ends the loop', async () => {
    const redex = new Redex();
    const tasks = await TasksAdapter.open(redex, ORIGIN);
    const seedSeq = tasks.create(1n, 'seed', 100n);
    await tasks.waitForSeq(seedSeq);

    const stream = await tasks.watch();
    const iter = stream[Symbol.asyncIterator]();

    // Await the watcher's initial emission directly — no timing guess.
    const first = await iter.next();
    expect(first.done).toBe(false);
    expect(first.value).toHaveLength(1);

    // Mutate and await the divergent emission.
    const seq = tasks.create(2n, 'second', 200n);
    await tasks.waitForSeq(seq);

    const second = await iter.next();
    expect(second.done).toBe(false);
    expect(second.value).toHaveLength(2);

    // `return()` releases the native iterator deterministically.
    await iter.return!();
  });
});

describe('TasksAdapter.snapshotAndWatch', () => {
  it('captures snapshot and streams subsequent deltas', async () => {
    // Baseline functional contract. Both skip(1) and skip_while would
    // pass this — the race regression test below is what actually
    // exercises the v2 fix.
    const redex = new Redex();
    const tasks = await TasksAdapter.open(redex, ORIGIN);
    const seq = tasks.create(1n, 'seed', 100n);
    await tasks.waitForSeq(seq);

    const { snapshot, updates } = await tasks.snapshotAndWatch();
    expect(snapshot.map((t) => t.id)).toEqual([1n]);

    const next = tasks.create(2n, 'post', 200n);
    await tasks.waitForSeq(next);

    const iter = updates[Symbol.asyncIterator]();
    const first = await iter.next();
    expect(first.done).toBe(false);
    const batch = first.value as Task[];
    expect(batch.map((t) => t.id).sort()).toEqual([1n, 2n]);
    await iter.return!();
  });

  it('regression: forwards divergent stream initial when snapshot races a mutation', async () => {
    // Mirrors
    // `test_regression_snapshot_and_watch_forwards_divergent_stream_initial`
    // in the core integration tests. Drives the race with concurrent
    // mutations; trials where the mutation already landed before the
    // snapshot read are skipped (nothing further to deliver). Under
    // the old skip(1) the race trials would hang because the
    // watcher's internal `last` already matched the post-mutation
    // state — hitting the timeout would fail the test.
    for (let trial = 0; trial < 20; trial++) {
      const redex = new Redex();
      const tasks = await TasksAdapter.open(redex, ORIGIN);
      const seq = tasks.create(1n, 'seed', 100n);
      await tasks.waitForSeq(seq);

      // Kick off mutation before the snapshot_and_watch call so it
      // races the two internal state reads.
      const mutation = (async () => {
        const s = tasks.create(2n, 'race', 200n);
        await tasks.waitForSeq(s);
      })();

      const { snapshot, updates } = await tasks.snapshotAndWatch();
      await mutation;

      // Mutation landed before the snapshot read: no further delta
      // coming. Skip this trial.
      if (snapshot.length === 2) continue;
      expect(snapshot.length, `trial ${trial}: initial should be [seed]`).toBe(1);

      const iter = updates[Symbol.asyncIterator]();
      const result = await Promise.race([
        iter.next(),
        new Promise<'timeout'>((r) => setTimeout(() => r('timeout'), 1000)),
      ]);
      if (result === 'timeout') {
        throw new Error(`trial ${trial}: stream must deliver post-snapshot state within timeout`);
      }
      expect(result.done).toBe(false);
      const batch = result.value as Task[];
      expect(batch.length, `trial ${trial}: stream must deliver state with both tasks`).toBe(2);
      await iter.return!();
    }
  });
});

describe('MemoriesAdapter.snapshotAndWatch', () => {
  it('regression: forwards divergent stream initial on race', async () => {
    // Mirror of the tasks race test for memories — same semantic, same
    // failure mode in the pre-fix skip(1) implementation.
    for (let trial = 0; trial < 20; trial++) {
      const redex = new Redex();
      const mem = await MemoriesAdapter.open(redex, ORIGIN);
      const seq = mem.store(1n, 'seed', ['t'], 'alice', 100n);
      await mem.waitForSeq(seq);

      const mutation = (async () => {
        const s = mem.store(2n, 'race', ['t'], 'alice', 200n);
        await mem.waitForSeq(s);
      })();

      const { snapshot, updates } = await mem.snapshotAndWatch();
      await mutation;

      if (snapshot.length === 2) continue;
      expect(snapshot.length, `trial ${trial}: initial should be [seed]`).toBe(1);

      const iter = updates[Symbol.asyncIterator]();
      const result = await Promise.race([
        iter.next(),
        new Promise<'timeout'>((r) => setTimeout(() => r('timeout'), 1000)),
      ]);
      if (result === 'timeout') {
        throw new Error(`trial ${trial}: stream must deliver post-snapshot state within timeout`);
      }
      expect(result.done).toBe(false);
      const batch = result.value;
      expect(batch!.length, `trial ${trial}: stream must deliver both memories`).toBe(2);
      await iter.return!();
    }
  });
});

describe('Error classification', () => {
  it('exposes CortexError class for instanceof checks', () => {
    // Constructor smoke: verifies the class is exported and throw/catch
    // paths behave as expected. The end-to-end "napi-thrown error
    // classified" path is already covered in
    // `bindings/node/test/errors.test.ts`; this guards the SDK re-export.
    const e = new CortexError('test');
    expect(e).toBeInstanceOf(CortexError);
    expect(e).toBeInstanceOf(Error);
    expect(e.name).toBe('CortexError');
  });
});

File: go/cortex_test.go
package net

import (
	"context"
	"fmt"
	"testing"
	"time"
)

const testOrigin = uint64(0xABCDEF01)

// ---------------------------------------------------------------------------
// Redex + RedexFile
// ---------------------------------------------------------------------------

func TestRedexFileAppendReadRange(t *testing.T) {
	r := NewRedex("")
	defer r.Free()

	f, err := r.OpenFile("go/test/basic", nil)
	if err != nil {
		t.Fatalf("open: %v", err)
	}
	defer f.Close()

	seq, err := f.Append([]byte("hello"))
	if err != nil {
		t.Fatalf("append: %v", err)
	}
	if seq != 0 {
		t.Fatalf("first seq should be 0, got %d", seq)
	}
	if f.Len() != 1 {
		t.Fatalf("Len should be 1, got %d", f.Len())
	}

	events, err := f.ReadRange(0, 10)
	if err != nil {
		t.Fatalf("read_range: %v", err)
	}
	if len(events) != 1 {
		t.Fatalf("expected 1 event, got %d", len(events))
	}
	if string(events[0].Payload) != "hello" {
		t.Fatalf("payload mismatch: %q", events[0].Payload)
	}
}

func TestRedexFileTail(t *testing.T) {
	r := NewRedex("")
	defer r.Free()

	f, err := r.OpenFile("go/test/tail", nil)
	if err != nil {
		t.Fatalf("open: %v", err)
	}
	defer f.Close()

	_, err = f.Append([]byte("early"))
	if err != nil {
		t.Fatalf("append: %v", err)
	}

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	events, errs, err := f.Tail(ctx, 0)
	if err != nil {
		t.Fatalf("tail: %v", err)
	}

	select {
	case ev := <-events:
		if string(ev.Payload) != "early" {
			t.Fatalf("backfill payload mismatch: %q", ev.Payload)
		}
	case <-time.After(1 * time.Second):
		t.Fatalf("timeout waiting for backfill")
	}

	if _, err := f.Append([]byte("live")); err != nil {
		t.Fatalf("append live: %v", err)
	}

	select {
	case ev := <-events:
		if string(ev.Payload) != "live" {
			t.Fatalf("live payload mismatch: %q", ev.Payload)
		}
	case err := <-errs:
		t.Fatalf("unexpected err from tail: %v", err)
	case <-time.After(1 * time.Second):
		t.Fatalf("timeout waiting for live")
	}

	cancel()
	// Drain to confirm the goroutine exits.
	drainTimeout := time.After(1 * time.Second)
	for {
		select {
		case _, ok := <-events:
			if !ok {
				return
			}
		case <-drainTimeout:
			t.Fatalf("tail goroutine didn't exit after context cancel")
		}
	}
}

// ---------------------------------------------------------------------------
// Tasks
// ---------------------------------------------------------------------------

func TestTasksCRUD(t *testing.T) {
	r := NewRedex("")
	defer r.Free()

	tasks, err := OpenTasks(r, testOrigin, false)
	if err != nil {
		t.Fatalf("open tasks: %v", err)
	}
	defer tasks.Close()

	if _, err := tasks.Create(1, "alpha", 100); err != nil {
		t.Fatalf("create: %v", err)
	}
	if _, err := tasks.Create(2, "beta", 200); err != nil {
		t.Fatalf("create: %v", err)
	}
	if _, err := tasks.Rename(1, "alpha-renamed", 300); err != nil {
		t.Fatalf("rename: %v", err)
	}
	seq, err := tasks.Complete(2, 400)
	if err != nil {
		t.Fatalf("complete: %v", err)
	}
	if err := tasks.WaitForSeq(seq, 2*time.Second); err != nil {
		t.Fatalf("wait_for_seq: %v", err)
	}

	list, err := tasks.List(nil)
	if err != nil {
		t.Fatalf("list: %v", err)
	}
	if len(list) != 2 {
		t.Fatalf("expected 2 tasks, got %d", len(list))
	}

	byID := map[uint64]Task{}
	for _, t := range list {
		byID[t.ID] = t
	}
	if byID[1].Title != "alpha-renamed" {
		t.Fatalf("task 1 title mismatch: %q", byID[1].Title)
	}
	if byID[2].Status != "completed" {
		t.Fatalf("task 2 should be completed, got %q", byID[2].Status)
	}
}

func TestTasksSnapshotAndWatchDeliversPostCallUpdates(t *testing.T) {
	r := NewRedex("")
	defer r.Free()
	tasks, err := OpenTasks(r, testOrigin, false)
	if err != nil {
		t.Fatalf("open: %v", err)
	}
	defer tasks.Close()

	seq, err := tasks.Create(1, "seed", 100)
	if err != nil {
		t.Fatalf("create: %v", err)
	}
	if err := tasks.WaitForSeq(seq, 2*time.Second); err != nil {
		t.Fatalf("wait_for_seq: %v", err)
	}

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	snapshot, updates, errs, err := tasks.SnapshotAndWatch(ctx, nil)
	if err != nil {
		t.Fatalf("snapshot_and_watch: %v", err)
	}
	if len(snapshot) != 1 {
		t.Fatalf("snapshot should be 1, got %d", len(snapshot))
	}

	seq2, err := tasks.Create(2, "post", 200)
	if err != nil {
		t.Fatalf("create post: %v", err)
	}
	if err := tasks.WaitForSeq(seq2, 2*time.Second); err != nil {
		t.Fatalf("wait_for_seq 2: %v", err)
	}

	select {
	case batch := <-updates:
		if len(batch) != 2 {
			t.Fatalf("expected 2 tasks in delta, got %d", len(batch))
		}
	case err := <-errs:
		t.Fatalf("unexpected err: %v", err)
	case <-time.After(2 * time.Second):
		t.Fatalf("timeout waiting for delta")
	}
}

// Regression test mirroring the Rust / TS / Python regression suites.
// Drives a concurrent mutation to race the snapshot / stream reads
// inside SnapshotAndWatch. Trials where the mutation landed before
// the snapshot read are skipped (no further delta to deliver). Under
// the pre-fix `skip(1)` the race trials would hang because the
// watcher's internal `last` already equalled the post-mutation state;
// here the timeout would fail the test.
func TestRegressionTasksSnapshotAndWatchForwardsDivergentInitial(t *testing.T) {
	for trial := 0; trial < 20; trial++ {
		r := NewRedex("")
		tasks, err := OpenTasks(r, testOrigin, false)
		if err != nil {
			t.Fatalf("trial %d: open: %v", trial, err)
		}

		seq, err := tasks.Create(1, "seed", 100)
		if err != nil {
			t.Fatalf("trial %d: create: %v", trial, err)
		}
		if err := tasks.WaitForSeq(seq, 2*time.Second); err != nil {
			t.Fatalf("trial %d: wait: %v", trial, err)
		}

		mutated := make(chan error, 1)
		go func() {
			s, err := tasks.Create(2, "race", 200)
			if err != nil {
				mutated <- fmt.Errorf("create: %w", err)
				return
			}
			if err := tasks.WaitForSeq(s, 2*time.Second); err != nil {
				mutated <- fmt.Errorf("wait_for_seq: %w", err)
				return
			}
			mutated <- nil
		}()

		ctx, cancel := context.WithCancel(context.Background())
		snapshot, updates, errs, err := tasks.SnapshotAndWatch(ctx, nil)
		if err != nil {
			cancel()
			t.Fatalf("trial %d: snapshot_and_watch: %v", trial, err)
		}
		if mErr := <-mutated; mErr != nil {
			// Fail fast with the real cause rather than waiting for a
			// misleading delta-timeout below.
			cancel()
			t.Fatalf("trial %d: mutator failed: %v", trial, mErr)
		}

		if len(snapshot) == 2 {
			cancel()
			tasks.Close()
			r.Free()
			continue
		}
		if len(snapshot) != 1 {
			cancel()
			t.Fatalf("trial %d: snapshot should be [seed], got %d", trial, len(snapshot))
		}

		select {
		case batch := <-updates:
			if len(batch) != 2 {
				cancel()
				t.Fatalf("trial %d: expected 2 in delta, got %d", trial, len(batch))
			}
		case err := <-errs:
			cancel()
			t.Fatalf("trial %d: watch err: %v", trial, err)
		case <-time.After(1 * time.Second):
			cancel()
			t.Fatalf("trial %d: timeout waiting for delta", trial)
		}
		cancel()
		tasks.Close()
		r.Free()
	}
}

// ---------------------------------------------------------------------------
// Memories
// ---------------------------------------------------------------------------

func TestMemoriesCRUD(t *testing.T) {
	r := NewRedex("")
	defer r.Free()
	mem, err := OpenMemories(r, testOrigin, false)
	if err != nil {
		t.Fatalf("open: %v", err)
	}
	defer mem.Close()

	if _, err := mem.Store(1, "hello", []string{"work"}, "alice", 100); err != nil {
		t.Fatalf("store: %v", err)
	}
	seq, err := mem.Pin(1, 110)
	if err != nil {
		t.Fatalf("pin: %v", err)
	}
	if err := mem.WaitForSeq(seq, 2*time.Second); err != nil {
		t.Fatalf("wait: %v", err)
	}

	list, err := mem.List(nil)
	if err != nil {
		t.Fatalf("list: %v", err)
	}
	if len(list) != 1 {
		t.Fatalf("expected 1 memory, got %d", len(list))
	}
	if !list[0].Pinned {
		t.Fatalf("memory should be pinned")
	}
}

func TestRegressionMemoriesSnapshotAndWatchForwardsDivergentInitial(t *testing.T) {
	for trial := 0; trial < 20; trial++ {
		r := NewRedex("")
		mem, err := OpenMemories(r, testOrigin, false)
		if err != nil {
			t.Fatalf("trial %d: open: %v", trial, err)
		}

		seq, err := mem.Store(1, "seed", []string{"t"}, "alice", 100)
		if err != nil {
			t.Fatalf("trial %d: store: %v", trial, err)
		}
		if err := mem.WaitForSeq(seq, 2*time.Second); err != nil {
			t.Fatalf("trial %d: wait: %v", trial, err)
		}

		mutated := make(chan error, 1)
		go func() {
			s, err := mem.Store(2, "race", []string{"t"}, "alice", 200)
			if err != nil {
				mutated <- fmt.Errorf("store: %w", err)
				return
			}
			if err := mem.WaitForSeq(s, 2*time.Second); err != nil {
				mutated <- fmt.Errorf("wait_for_seq: %w", err)
				return
			}
			mutated <- nil
		}()

		ctx, cancel := context.WithCancel(context.Background())
		snapshot, updates, errs, err := mem.SnapshotAndWatch(ctx, nil)
		if err != nil {
			cancel()
			t.Fatalf("trial %d: snapshot_and_watch: %v", trial, err)
		}
		if mErr := <-mutated; mErr != nil {
			cancel()
			t.Fatalf("trial %d: mutator failed: %v", trial, mErr)
		}

		if len(snapshot) == 2 {
			cancel()
			mem.Close()
			r.Free()
			continue
		}
		if len(snapshot) != 1 {
			cancel()
			t.Fatalf("trial %d: snapshot should be [seed], got %d", trial, len(snapshot))
		}

		select {
		case batch := <-updates:
			if len(batch) != 2 {
				cancel()
				t.Fatalf("trial %d: expected 2 in delta, got %d", trial, len(batch))
			}
		case err := <-errs:
			cancel()
			t.Fatalf("trial %d: watch err: %v", trial, err)
		case <-time.After(1 * time.Second):
			cancel()
			t.Fatalf("trial %d: timeout waiting for delta", trial)
		}
		cancel()
		mem.Close()
		r.Free()
	}
}
